src/genesis/mod.rs

```diff
@@ -139,42 +142,65 @@ where
 /// header.
 async fn load_and_retrieve_header<DB, R>(
     store: DB,
-    mut reader: R,
+    reader: R,
     skip_load: bool,
 ) -> anyhow::Result<(Vec<Cid>, Option<usize>)>
 where
     DB: Blockstore + Send + Sync + 'static,
     R: AsyncRead + Send + Unpin,
 {
     let result = if skip_load {
-        (CarReader::new(&mut reader).await?.header.roots, None)
+        (CarReader::new(reader).await?.header.roots, None)
     } else {
-        let (roots, n_records) = forest_load_car(store, &mut reader).await?;
+        let (roots, n_records) = forest_load_car(store, reader).await?;
         (roots, Some(n_records))
     };
 
     Ok(result)
 }
 
+fn car_stream<R: futures::AsyncRead + Send + Unpin>(
+    mut reader: CarReader<R>,
+) -> impl Stream<Item = anyhow::Result<fvm_ipld_car::Block>> {
+    try_stream! {
+        while let Some(block) = reader.next_block().await? {
+            yield block;
+        }
+    }
+}
+
 pub async fn forest_load_car<DB, R>(store: DB, reader: R) -> anyhow::Result<(Vec<Cid>, usize)>
 where
     R: futures::AsyncRead + Send + Unpin,
     DB: Blockstore + Send + Sync + 'static,
 {
-    // 1GB
-    const BUFFER_CAPCITY_BYTES: usize = 1024 * 1024 * 1024;
+    let db = Arc::new(store);
 
-    let (tx, rx) = flume::bounded(100);
-    let write_task =
-        tokio::spawn(async move { store.buffered_write(rx, BUFFER_CAPCITY_BYTES).await });
-    let mut car_reader = CarReader::new(reader).await?;
+    let car_reader = CarReader::new(reader).await?;
+    let roots = car_reader.header.roots.clone();
     let mut n_records = 0;
-    while let Some(block) = car_reader.next_block().await? {
-        debug!("Importing block: {}", block.cid);
-        n_records += 1;
-        tx.send_async((block.cid, block.data)).await?;
-    }
-    drop(tx);
-    write_task.await??;
-    Ok((car_reader.header.roots, n_records))
+
+    let sink = |block: Vec<fvm_ipld_car::Block>| {
```

Comment:
Why introduce a `sink` here? You can just map (concurrently!) into the blockstore:
https://github.com/ChainSafe/forest/blob/cd8517e84114b2bc74cad424776898cb544b3ba5/src/car_stream.rs#L34-L37

Reply:
_Or_ create a struct `CarSink<BlockstoreT>` and `impl Sink for CarSink where BlockstoreT: Blockstore`

Reply:
Concurrent mapping doesn't work as one might have hoped with non-asynchronous code: It tries to run the code concurrently rather than in parallel, but `put_many_keyed` cannot run concurrently. In this case, `for_each_concurrent` behave exactly as `for_each`, and you don't get the desired buffering behavior.

Refactoring into a `CarSink` sounds like a good idea.